File: src/game.zig
const std = @import("std");
const c = @import("c.zig");
const fmt = std.fmt;
const Allocator = std.mem.Allocator;
const zm = @import("zmath");
const SpriteRenderer = @import("sprite_renderer.zig").SpriteRenderer;
const TextRenderer = @import("text_renderer.zig").TextRenderer;
const Camera = @import("camera.zig").Camera;
const Texture = @import("texture.zig").Texture;
const Dungeon = @import("dungeon.zig").Dungeon;
const Grid = @import("grid.zig").Grid;
const Hero = @import("hero.zig").Hero;
const Enemy = @import("enemy.zig").Enemy;
const Clock = @import("clock.zig").Clock;

const SCREEN_WIDTH: i32 = 800;
const SCREEN_HEIGHT: i32 = 600;

pub const Game = struct {
    const Self = @This();
    shouldQuit: bool,
    camera: Camera,
    spriteRenderer: SpriteRenderer,
    textRenderer: TextRenderer,
    dungeon: Dungeon,
    grid: Grid,
    clock: Clock,

    pub fn init(allocator: *Allocator) !Self {
        var spriteRenderer = SpriteRenderer.init(allocator);
        var textRenderer = try TextRenderer.init(allocator);

        spriteRenderer.shader.use();
        var projection = zm.orthographicRh(800.0, 600.0, -10.0, 10.0);
        spriteRenderer.shader.set_mat4("projection", projection);

        var dungeon = try Dungeon.init(allocator, 50);
        var grid = Grid.fromSpriteSize(zm.f32x4(32.0, 32.0, 0.0, 0.0));

        return Self{
            .camera = Camera.init(zm.f32x4(0.0, 0.0, 0.0, 0.0)),
            .spriteRenderer = spriteRenderer,
            .shouldQuit = false,
            .textRenderer = textRenderer,
            .dungeon = dungeon,
            .grid = grid,
            .clock = Clock.init(),
        };
    }

    pub fn deinit(self: *Self) void {
        self.textRenderer.deinit();
        self.spriteRenderer.deinit();
        self.dungeon.deinit();
    }

    pub fn processEvent(self: *Self, e: c.SDL_Event) void {
        self.dungeon.processEvent(e);
    }

    pub fn update(self: *Self) void {
        var spriteSize = self.grid.spriteSize();
        spriteSize[0] /= 2.0;
        spriteSize[1] /= 2.0;
        self.camera.focusOn(self.grid.toScreenPosition(self.dungeon.hero.gridPosition) + spriteSize);
        self.spriteRenderer.shader.use();
        self.spriteRenderer.shader.set_mat4("view", self.camera.viewMatrix());
    }

    pub fn draw(self: *Self) !void {
        self.dungeon.draw(&self.spriteRenderer, &self.grid);

        var buf: [15]u8 = undefined;
        const result = try fmt.bufPrint(&buf, "Frame: {d}ms", .{self.clock.deltaTime()});

        self.textRenderer.draw(result, -SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2 - 20, 0.4, zm.f32x4(1.0, 1.0, 1.0, 0.0));
    }
};

File: src/game.cpp
#include <glm/gtc/matrix_transform.hpp>
#include "game.h"
#include "renderer_repo.h"

#define UNUSED(X) (void)(X)

Game::Game(unsigned int viewport_width, unsigned int viewport_height) {
	width = viewport_width;
	height = viewport_height;
	keys = (bool *)malloc(sizeof(bool) * 1024);
	state = GAME_ACTIVE;
	text_renderer = NULL;
	dungeon = NULL;
	should_quit = false;

	input_map = std::map<SDL_Keycode, Input> {
		{ SDLK_UP, Input::ArrowUp },
		{ SDLK_LEFT, Input::ArrowLeft },
		{ SDLK_DOWN, Input::ArrowDown },
		{ SDLK_RIGHT, Input::ArrowRight },
		{ SDLK_RETURN, Input::Enter },
		{ SDLK_x, Input::X }
	};
}

Game::~Game() {
	delete text_renderer;
	delete dungeon;
	free(keys);
}

void Game::init() {
	Shader *shader = new Shader();
	shader->create((char *)"shaders/default.vert", (char *)"shaders/default.frag");

	Shader *blink_shader = new Shader();
	blink_shader->create((char *)"shaders/default.vert", (char *)"shaders/blink.frag");

	glm::mat4 projection = glm::ortho(0.0f, (float)width, (float)height, 0.0f, -10.0f, 10.0f);

	renderer_repo["default"] = new SpriteRenderer(shader);
	renderer_repo["blink"] = new SpriteRenderer(blink_shader);
	printf("Setting renderers...\n");

	shader->use();
	shader->set_int((char *)"image", 0);
	shader->set_mat4((char *)"projection", projection);

	blink_shader->use();
	blink_shader->set_int((char *)"image", 0);
	blink_shader->set_mat4((char *)"projection", projection);

	text_renderer = new TextRenderer(projection);

	dungeon = new Dungeon(50);
}

bool Game::handle_input(SDL_Event e) {
	Input input = Input::Unknown;
	bool handled = false;

	// TODO: Maybe this translation should be done on App level
	switch(e.type) {
	case SDL_KEYDOWN:
		input = input_map[e.key.keysym.sym];
		handled = dungeon->handle_input(input);
		break;
	default:
		break;
	}

	return handled;
}

void Game::update(float delta_time) {
	if (dungeon->game_over_action == GameOverAction::Quit) {
		should_quit = true;
	}
	else if (dungeon->game_over_action == GameOverAction::Restart) {
		delete dungeon;
		dungeon = new Dungeon(50);
	}
	else {
		dungeon->update(delta_time);

		auto renderer = renderer_repo["default"];
		auto shader = renderer->shader;
		auto b_renderer = renderer_repo["blink"];
		auto blink_shader = b_renderer->shader;
		shader->use();

		auto view = dungeon->camera->view_matrix();
		shader->set_mat4((char *)"view", view);

		blink_shader->use();
		blink_shader->set_mat4((char *)"view", view);
	}
}

void Game::render() {
	dungeon->render(text_renderer);
}
